junit-engine-api/src/main/java/org/junit/gen5/engine/AbstractTestDescriptor.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.engine;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public abstract class AbstractTestDescriptor implements TestDescriptor {
+
+	private final String uniqueId;
+	private TestDescriptor parent;
+	private final Set<TestDescriptor> children = new HashSet<>();
+
+	protected AbstractTestDescriptor(String uniqueId) {
+		this.uniqueId = uniqueId;
+	}
+
+	@Override
+	public boolean equals(Object other) {
+		if (other == null)
+			return false;
+		if (this.getClass() != other.getClass())
+			return false;
+		TestDescriptor otherDescriptor = (TestDescriptor) other;
+		return this.getUniqueId().equals(otherDescriptor.getUniqueId());
+	}
+
+	@Override
+	public int hashCode() {
```

Comment:
My point: Without a known usage pattern nobody can tell what efficiency means. Calculating a hash code also takes time. If you're not reasonably certain that it will prevent collisions in most cases, calculating the hash at all is inefficient. 

> Am 07.11.2015 um 08:50 schrieb Kevin Cooney <notifications@github.com>:
> 
> In junit-engine-api/src/main/java/org/junit/gen5/engine/AbstractTestDescriptor.java:
> 
> > +	protected AbstractTestDescriptor(String uniqueId) {
> > +		this.uniqueId = uniqueId;
> > +	}
> > +
> > +	@Override
> > +	public boolean equals(Object other) {
> > +		if (other == null)
> > +			return false;
> > +		if (this.getClass() != other.getClass())
> > +			return false;
> > +		TestDescriptor otherDescriptor = (TestDescriptor) other;
> > +		return this.getUniqueId().equals(otherDescriptor.getUniqueId());
> > +	}
> > +
> > +	@Override
> > +	public int hashCode() {
> Yea, hashing collections fall back to equals, but 1) that could be inefficient, and 2) other code using hashCode() could be even less efficient with many collisions.
> 
> —
> Reply to this email directly or view it on GitHub.
>